### PR TITLE
IBP 3798 Advance and Cross list refactor

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/java/org/ibp/api/domain/study/StudyGermplasmSourceTable.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package org.ibp.api.domain.study;
+
+import org.generationcp.middleware.service.api.study.germplasm.source.StudyGermplasmSourceDto;
+import org.pojomatic.Pojomatic;
+import org.pojomatic.annotations.AutoProperty;
+
+import java.util.List;
+
+@AutoProperty
+public class StudyGermplasmSourceTable {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/domain/study/StudyGermplasmSourceTable.java]: Rename to GermplasmStudySourceTable

File: src/main/java/org/ibp/api/java/impl/middleware/study/StudyGermplasmSourceServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import org.ibp.api.java.study.StudyGermplasmSourceService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.annotation.Resource;
+import java.util.List;
+
+@Service
+@Transactional
+public class StudyGermplasmSourceServiceImpl implements StudyGermplasmSourceService {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/study/StudyGermplasmSourceServiceImpl.java]: Rename to GermplasmStudySourceServiceXXX

File: src/main/java/org/ibp/api/java/impl/middleware/study/StudyGermplasmSourceServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			.getStudyGermplasmSourceList(studyGermplasmSourceRequest);
+
+		for (final StudyGermplasmSourceDto studyGermplasmSourceDto : studyGermplasmSourceDtoList) {
+			studyGermplasmSourceDto.setCross(this.pedigreeService.getCrossExpansion(studyGermplasmSourceDto.getGid(), this.crossExpansionProperties));
+		}
+		return studyGermplasmSourceDtoList;
+	}
+
+	@Override
+	public long countStudyGermplasmSourceList(final StudyGermplasmSourceRequest studyGermplasmSourceRequest) {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/study/StudyGermplasmSourceServiceImpl.java]: Why do we need 2 methods having the same input: countStudyGermplasmSourceList and  countFilteredStudyGermplasmSourceList

File: src/main/java/org/ibp/api/domain/study/StudyGermplasmSourceTable.java (hunk 1 above)
Comment (abatac) [reply src/main/java/org/ibp/api/domain/study/StudyGermplasmSourceTable.java]: Done

File: src/main/java/org/ibp/api/java/impl/middleware/study/StudyGermplasmSourceServiceImpl.java (hunk 2 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/impl/middleware/study/StudyGermplasmSourceServiceImpl.java]: Done

File: src/main/java/org/ibp/api/java/impl/middleware/study/StudyGermplasmSourceServiceImpl.java (hunk 3 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/impl/middleware/study/StudyGermplasmSourceServiceImpl.java]: countStudyGermplasmSourceList gets the total number of all germplasm created in the study. countFilteredStudyGermplasmSourceList gets the number of germplasm created in the study with filter being applied.

We don't actually show the count of filtered result in the UI. But I implemented this to satisfy the standard SearchSpec interface.

File: src/main/java/org/ibp/api/java/impl/middleware/study/GermplasmStudySourceServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	@Override
+	public List<GermplasmStudySourceDto> getGermplasmStudySourceList(final GermplasmStudySourceRequest germplasmStudySourceRequest) {
+
+		this.studyValidator.validate(germplasmStudySourceRequest.getStudyId(), false);
+
+		final List<GermplasmStudySourceDto> germplasmStudySourceDtoList = this.germplasmStudySourceMiddlewareService
+			.getGermplasmStudySourceList(germplasmStudySourceRequest);
+
+		for (final GermplasmStudySourceDto germplasmStudySourceDto : germplasmStudySourceDtoList) {
+			germplasmStudySourceDto.setCross(this.pedigreeService.getCrossExpansion(germplasmStudySourceDto.getGid(), this.crossExpansionProperties));
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/study/GermplasmStudySourceServiceImpl.java:41]: @darla-leafnode will this perform well? we had some issues calling the cross expansion before.

File: src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package org.ibp.api.java.study;
+
+import org.generationcp.middleware.service.api.study.germplasm.source.GermplasmStudySourceDto;
+import org.generationcp.middleware.service.api.study.germplasm.source.GermplasmStudySourceRequest;
+
+import java.util.List;
+
+public interface GermplasmStudySourceService {
+
+	public List<GermplasmStudySourceDto> getGermplasmStudySourceList(final GermplasmStudySourceRequest germplasmStudySourceRequest);
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java:10]: remove final

File: src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+	public long countGermplasmStudySourceList(final GermplasmStudySourceRequest germplasmStudySourceRequest);
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java:12]: remove final

File: src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+	public long countFilteredGermplasmStudySourceList(final GermplasmStudySourceRequest germplasmStudySourceRequest);
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java:14]: remove final

File: src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java (hunk 5 above)
Comment (clarysabel) [reply src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java:10]: same comment than in MIddleware, using List at the end will make it more difficult to understand, specially the count. I will prefer to use plural for the name of these functions

File: src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	@ApiOperation(value = "It will retrieve all generated germplasm in a study",
+		notes = "It will retrieve all generated germplasm in a study")
+	@PreAuthorize("hasAnyAuthority('ADMIN','BREEDING_ACTIVITIES','MANAGE_STUDIES')")
+	@RequestMapping(value = "/{cropname}/programs/{programUUID}/studies/{studyId}/germplasm-sources/table", method = RequestMethod.POST)
+	@ResponseBody
+	public ResponseEntity<GermplasmStudySourceTable> getGermplasmStudySourceTable(final @PathVariable String cropname,
+		@PathVariable final String programUUID,
+		@PathVariable final Integer studyId, @RequestBody final GermplasmStudySourceRequest germplasmStudySourceRequest) {
+
+		Preconditions.checkNotNull(germplasmStudySourceRequest, "params cannot be null");
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java:45]: we are avoiding using Preconditions for validations that will end at the client side, Please use BaseValidator.checkNotNull instead

File: src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+		final SortedPageRequest sortedRequest = germplasmStudySourceRequest.getSortedRequest();
+		Preconditions.checkNotNull(sortedRequest, "sortedRequest inside params cannot be null");
+		final String sortOrder = germplasmStudySourceRequest.getSortedRequest().getSortOrder();
+		final boolean isSortOrderValid = "ASC".equalsIgnoreCase(sortOrder) || "DESC".equalsIgnoreCase(sortOrder) || StringUtils.isEmpty(sortOrder);
+		Preconditions.checkArgument(isSortOrderValid, "sortOrder should be either ASC or DESC");
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java:50]: use BaseValidator

File: src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+	@PreAuthorize("hasAnyAuthority('ADMIN','BREEDING_ACTIVITIES','MANAGE_STUDIES')")
+	@RequestMapping(value = "/{cropname}/programs/{programUUID}/studies/{studyId}/germplasm-sources/table", method = RequestMethod.POST)
+	@ResponseBody
+	public ResponseEntity<GermplasmStudySourceTable> getGermplasmStudySourceTable(final @PathVariable String cropname,
+		@PathVariable final String programUUID,
+		@PathVariable final Integer studyId, @RequestBody final GermplasmStudySourceRequest germplasmStudySourceRequest) {
+
+		Preconditions.checkNotNull(germplasmStudySourceRequest, "params cannot be null");
+		final SortedPageRequest sortedRequest = germplasmStudySourceRequest.getSortedRequest();
+		Preconditions.checkNotNull(sortedRequest, "sortedRequest inside params cannot be null");
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java:47]: use BaseValidator

File: src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java (hunk 5 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java:10]: Done

File: src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java (hunk 6 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java:12]: Done

File: src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java (hunk 7 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java:14]: Done

File: src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java (hunk 8 above)
Comment (abatac) [reply src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java:45]: Done

File: src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java (hunk 9 above)
Comment (abatac) [reply src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java:50]: Done

File: src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java (hunk 10 above)
Comment (abatac) [reply src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java:47]: Done

File: src/main/java/org/ibp/api/java/impl/middleware/study/GermplasmStudySourceServiceImpl.java (hunk 4 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/impl/middleware/study/GermplasmStudySourceServiceImpl.java:41]: In our testing, this performs fast as we only process the records on the current page (100 max). But we can further improve this by using this.pedigreeService.getCrossExpansions to minimize the expense of calling crossExpansionProperties.getCropGenerationLevel multiple times.

Comment (abatac) [reply src/main/java/org/ibp/api/java/impl/middleware/study/GermplasmStudySourceServiceImpl.java:41]: Done

**Comment by abatac** [comment]
Hi @clarysabel, I'm done with the requested changes. Please check.

File: src/main/java/org/ibp/api/java/impl/middleware/study/StudyServiceImpl.java [hunk 11]
```diff
@@ -400,7 +385,7 @@ public boolean studyHasGivenDatasetType(Integer studyId, Integer datasetTypeId)
 	}
 
 	@Override
-	public boolean hasAdvancedOrCrossesList(int studyId) {
-		return this.middlewareStudyService.hasAdvancedOrCrossesList(studyId);
+	public boolean hasCrossesOrSelections(int studyId) {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/study/StudyServiceImpl.java:388]: Add final to parameters here and `studyHasGivenDatasetType` method

File: src/main/java/org/ibp/api/java/study/StudyService.java [hunk 12]
```diff
@@ -72,5 +70,5 @@
 
 	boolean studyHasGivenDatasetType(Integer studyId, Integer datasetTypeId);
 
-	boolean hasAdvancedOrCrossesList(final int studyId);
+	boolean hasCrossesOrSelections(final int studyId);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/study/StudyService.java:73]: Remove `final` in this class

File: src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
+package org.ibp.api.rest.study;
+
+import com.google.common.base.Preconditions;
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java:3]: Remove unused import

File: src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+@RequestMapping("/crops")
+public class GermplasmStudySourceResource {
+
+	@Resource
+	private GermplasmStudySourceService germplasmStudySourceService;
+
+	@ApiOperation(value = "It will retrieve all generated germplasm in a study",
+		notes = "It will retrieve all generated germplasm in a study")
+	@PreAuthorize("hasAnyAuthority('ADMIN','BREEDING_ACTIVITIES','MANAGE_STUDIES')")
+	@RequestMapping(value = "/{cropname}/programs/{programUUID}/studies/{studyId}/germplasm-sources/table", method = RequestMethod.POST)
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java:40]: As a rule, we add the following permissions for GET calls for study:
'INFORMATION_MANAGEMENT', 'BROWSE_STUDIES'

File: src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+	@ResponseBody
+	public ResponseEntity<GermplasmStudySourceTable> getGermplasmStudySourceTable(final @PathVariable String cropname,
+		@PathVariable final String programUUID,
+		@PathVariable final Integer studyId, @RequestBody final GermplasmStudySourceSearchRequest germplasmStudySourceSearchRequest) {
+
+		BaseValidator.checkNotNull(germplasmStudySourceSearchRequest, "params cannot be null");
+		final SortedPageRequest sortedRequest = germplasmStudySourceSearchRequest.getSortedRequest();
+		BaseValidator.checkNotNull(sortedRequest, "sortedRequest inside params cannot be null");
+		final String sortOrder = germplasmStudySourceSearchRequest.getSortedRequest().getSortOrder();
+		final boolean isSortOrderValid = "ASC".equalsIgnoreCase(sortOrder) || "DESC".equalsIgnoreCase(sortOrder) || StringUtils.isEmpty(sortOrder);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/study/GermplasmStudySourceResource.java:50]: Ideally, we want a common validator that validates value for sortOrder but https://ibplatform.atlassian.net/browse/IBP-3899 will be addressed on 15.4

**Pull request merged**